### PR TITLE
[DC] Add ops and types

### PR DESCRIPTION
**Pull request by mortbopet**

### CODE REVIEW

File: include/circt/Dialect/DC/DCOps.td [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+      IsolatedFromAbove,
+      Symbol,
+      FunctionOpInterface,
+      RegionKindInterface,
+      CallableOpInterface,
+  ]> {
+  // I really don't want to have to add yet another container operation. However,
+  // where else do we place DC ops?
+  // We have to place them in a graph region operation, so func.func isn't
+  // applicable. We're not committed to being in `hw` yet, so hw.module doesn't
```
Comment (teqdruid) [inline include/circt/Dialect/DC/DCOps.td:43]: If I understand the DC dialect properly, these ops can go in any graph container block. You just have to select one (or several) purely for testing and only in the test files. In which case `hw.module` is appropriate. The `msft.module` op could also be used.

So I'm going to ignore this and the next few ops.

**Review by teqdruid — COMMENTED** [review]
I haven't really had time to grok this, but had a comment.

File: lib/Dialect/DC/DCOps.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  if (!vt || vt.getInnerTypes().size() != 1)
+    return false;
+
+  auto innerWidth = vt.getInnerTypes()[0].getIntOrFloatBitWidth();
+  return innerWidth == 1;
+}
+
+namespace circt {
+namespace dc {
+#include "circt/Dialect/DC/DCCanonicalization.h.inc"
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:31]: I don't understand why this is necessary.

File: lib/Dialect/DC/DCOps.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+FunctionType CallOp::getCalleeType() {
+  return FunctionType::get(getContext(), getOperandTypes(), getResultTypes());
+}
+
+// =============================================================================
+// JoinOp
+// =============================================================================
+
+void JoinOp::getCanonicalizationPatterns(RewritePatternSet &results,
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:124]: Why? Presumably you'll be filling this in in a subsequent PR?

File: lib/Dialect/DC/DCOps.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  auto innerWidth = vt.getInnerTypes()[0].getIntOrFloatBitWidth();
+  return innerWidth == 1;
+}
+
+namespace circt {
+namespace dc {
+#include "circt/Dialect/DC/DCCanonicalization.h.inc"
+
+// =============================================================================
+// FuncOp
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:34]: I'm just going to ignore these under the assumption that they'll go away before a merge.

File: lib/Dialect/DC/DCOps.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  llvm::SMLoc allOperandLoc = parser.getCurrentLocation();
+  size_t size = 0;
+  if (parseIntInSquareBrackets(parser, size))
+    return failure();
+
+  if (size == 0)
+    return parser.emitError(parser.getNameLoc(),
+                            "fork size must be greater than 0");
+
+  if (parser.parseOperandList(operands) ||
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:149]: I thought that `fork` only ever had one operand? If so, `parseOperand` would be more appropriate so you don't accept `%a, %b`.

File: lib/Dialect/DC/DCOps.cpp [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+      parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  auto tt = dc::TokenType::get(parser.getContext());
+  llvm::SmallVector<Type> operandTypes{tt};
+  SmallVector<Type, 1> resultTypes{size, tt};
+  result.addTypes(resultTypes);
+  if (parser.resolveOperands(operands, operandTypes, allOperandLoc,
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:157]: Likewise, use `parser.resolveOperand` (non-plural) if my understanding about fork is correct.

**Comment by teqdruid** [comment]
Also, tests? Round tripping for the custom parsers would be appropriate. Also, I would like to see some round trip and/or error tests for things not guaranteed correct by the ODS assemblyFormat (e.g. type constraints).

File: include/circt/Dialect/DC/DCOps.td [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+  let arguments = (ins Variadic<AnyType>:$operands);
+  let assemblyFormat = "($operands^ `:` type($operands))? attr-dict";
+}
+
+def BufferOp : DCOp<"buffer",
+    [SameTypeConstraint<"input", "output">]> {
+  let summary = "Buffer operation";
+  let description = [{
+    The buffer operation may buffer a control-side token against an arbitrary amount
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/DC/DCOps.td:203]: Yea, that is essentially it. I think i was over-elaborating here a bit to give the notion that buffers can be used to delay data-side values wherein that delay is control-sensitive (as opposed to "just" a register). I'll reformulate.

File: include/circt/Dialect/DC/DCOps.td [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    ValueOrTokenType:$input,
+    ConfinedAttr<I64Attr, [IntMinValue<1>]>:$size,
+    OptionalAttr<AnyAttr>:$initValues
+  );
+  let results = (outs ValueOrTokenType:$output);
+
+  let assemblyFormat = "$input `[` $size `]` attr-dict `:` type($input)";
+  let builders = [OpBuilder<(
+    ins "Value":$input, "size_t":$size), [{
+        build(odsBuilder, odsState, input.getType(), input, odsBuilder.getI64IntegerAttr(size), {});
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/DC/DCOps.td:222]: 😉 fixed!

File: lib/Dialect/DC/DCOps.cpp (hunk 3 above)
Comment (mortbopet) [reply lib/Dialect/DC/DCOps.cpp:124]: Yep! Added them here to keep the "hasCanonicalizer" in the ODS in this PR.

File: lib/Dialect/DC/DCOps.cpp (hunk 2 above)
Comment (mortbopet) [reply lib/Dialect/DC/DCOps.cpp:31]: Not necessary for this PR - if we're real nit-ty, i could remove it in this PR and move it to the canonicalization PR 👀.

File: include/circt/Dialect/DC/DCOps.td [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+    let builders = [OpBuilder<(ins), [{
+            build(odsBuilder, odsState, odsBuilder.getType<dc::TokenType>());
+        }]>];
+}
+
+def PackOp : DCOp<"pack"> {
+    let summary = "Pack operation";
+    let description = [{
+        An operation which packs together a !dc.token value with some other
+        value - mostly used to facilitate conversion, given the lack of
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/DC/DCOps.td:358]: True - stale comment that originates from the early prototyping of the dialect.

**Comment by mortbopet** [comment]
tests have now been added.

File: lib/Dialect/DC/DCOps.cpp [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    return parser.emitError(parser.getNameLoc(),
+                            "fork size must be greater than 0");
+
+  if (parser.parseOperand(operand) ||
+      parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  auto tt = dc::TokenType::get(parser.getContext());
+  llvm::SmallVector<Type> operandTypes{tt};
+  SmallVector<Type, 1> resultTypes{size, tt};
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:67]: Nit: is 1 really a good assumption? A fork with one output would be kind of silly. I think 8 or just leaving it off would be more appropriate.

File: lib/Dialect/DC/DCOps.cpp [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+// =============================================================================
+
+FailureOr<SmallVector<int64_t>> BufferOp::getInitValueArray() {
+  assert(getInitValues() && "initValues attribute not set");
+  SmallVector<int64_t> values;
+  for (auto value : getInitValuesAttr()) {
+    if (auto iValue = value.dyn_cast<IntegerAttr>()) {
+      values.push_back(iValue.getValue().getSExtValue());
+    } else {
+      return emitError() << "initValues attribute must be an array of integers";
```
Comment (teqdruid) [inline lib/Dialect/DC/DCOps.cpp:172]: Better, but in what case is this assumption reasonable? I suspect that outside of testing, this case would be rare.

**Review by teqdruid — APPROVED** [review]
LGTM! Thanks for the revisions.

**Pull request merged**